File: src/bmsx/render/vdp/readback.ts
import { $ } from '../../core/engine';

export function readVdpTextureRegion(textureKey: string, x: number, y: number, width: number, height: number, out?: Uint8Array): Uint8Array {
	return $.view.backend.readTextureRegion($.texmanager.getTextureByUri(textureKey), x, y, width, height, out);
}

export function vdpTextureReadbackUsesCpuCopy(): boolean {
	return $.view.backend.type === 'webgpu';
}

File: src/bmsx_cpp/render/vdp/readback.cpp
#include "render/vdp/readback.h"

#include "core/engine.h"
#include "render/texture_manager.h"

namespace bmsx {

void readVdpTextureRegion(const std::string& textureKey, u8* out, i32 width, i32 height, i32 x, i32 y) {
	auto* texmanager = EngineCore::instance().texmanager();
	texmanager->backend()->readTextureRegion(texmanager->getTextureByUri(textureKey), out, width, height, x, y, {});
}

} // namespace bmsx
